sql/2020/01_CSS/vendor_prefix_summary.sql

```diff
@@ -0,0 +1,113 @@
+#standardSQL
+CREATE TEMPORARY FUNCTION getPrefixStats(css STRING) RETURNS ARRAY<STRING> LANGUAGE js AS '''
+try {
+	function compute() {
+		let ret = {
+			total: {},
+			pseudo_classes: {},
+			pseudo_elements: {},
+			properties: {},
+			functions: {},
+			keywords: {},
+			media: {}
+		};
+
+		ret.total = Object.fromEntries(Object.keys(ret).map(k => [k, 0]));
+
+		walkRules(ast, rule => {
+			// Prefixed pseudos
+			if (rule.selectors) {
+				let pseudos = rule.selectors.flatMap(r => r.match(/::?-[a-z]+-[\\w-]+/g) || []);
+
+				for (let pseudo of pseudos) {
+					let type = "pseudo_" + (pseudo.indexOf("::") === 0? "elements" : "classes");
+					incrementByKey(ret[type], pseudo);
+					ret.total[type]++;
+				}
+			}
+
+			if (rule.declarations) {
+				walkDeclarations(rule, ({property, value}) => {
```

Comment:
The `// -prefix-function()` section is causing this query to timeout. I think a value check might prevent that. The query takes a very long time to run and I'm logging off for the night but wanted to give that suggestion for you to see in the morning.
```suggestion
				walkDeclarations(rule, ({property, value}) => {
				if (value.length > 1000) {
					return;
				}
```

Reply:
Still no results. Also note that this is happening with the `parsed_css_1k` table, so I think it may actually be returning no results due to a bug with the logic.

Reply:
Confirmed that the suggested change produces results:


client | prop | pages | freq | total | pct
-- | -- | -- | -- | -- | --
desktop | total | 5036308 | 2934010328 | 5868020656 | 0.5
mobile | total | 5779675 | 3899424875 | 7798849750 | 0.5
mobile | properties | 5760913 | 3265650025 | 7798849750 | 0.4187348301
desktop | properties | 5015245 | 2453285776 | 5868020656 | 0.4180772223
mobile | keywords | 4884489 | 264043378 | 7798849750 | 0.03385670791
desktop | keywords | 4116709 | 184325267 | 5868020656 | 0.03141182995
desktop | functions | 3540529 | 141897399 | 5868020656 | 0.02418147572
mobile | functions | 4155633 | 172619077 | 7798849750 | 0.02213391494
desktop | pseudo_elements | 4228723 | 97160938 | 5868020656 | 0.01655770211
mobile | pseudo_elements | 4860030 | 127033773 | 7798849750 | 0.01628878323
desktop | pseudo_classes | 3385086 | 37292161 | 5868020656 | 0.006355151624
mobile | pseudo_classes | 3843642 | 44044055 | 7798849750 | 0.005647506544
desktop | media | 2569576 | 20048787 | 5868020656 | 0.003416618341
mobile | media | 3316548 | 26034567 | 7798849750 | 0.003338257286

Reply:
Here's the full query that I ran to get those results:

```sql
#standardSQL
CREATE TEMPORARY FUNCTION getPrefixStats(css STRING) RETURNS ARRAY<STRING> LANGUAGE js AS '''
try {
  function compute() {
    let ret = {
      total: {},
      pseudo_classes: {},
      pseudo_elements: {},
      properties: {},
      functions: {},
      keywords: {},
      media: {}
    };

    ret.total = Object.fromEntries(Object.keys(ret).map(k => [k, 0]));

    walkRules(ast, rule => {
      // Prefixed pseudos
      if (rule.selectors) {
        let pseudos = rule.selectors.flatMap(r => r.match(/::?-[a-z]+-[\\w-]+/g) || []);

        for (let pseudo of pseudos) {
          let type = "pseudo_" + (pseudo.indexOf("::") === 0? "elements" : "classes");
          incrementByKey(ret[type], pseudo);
          ret.total[type]++;
        }
      }

      if (rule.declarations) {
        walkDeclarations(rule, ({property, value}) => {
          if (value.length > 1000) {
            return;
          }

          // Prefixed properties
          if (/^-[a-z]+-.+/.test(property)) {
            incrementByKey(ret.properties, property);
            ret.total.properties++;
          }

          // -prefix-function()
          for (let call of extractFunctionCalls(value, {names: /^-[a-z]+-.+/})) {
            incrementByKey(ret.functions, call.name);
            ret.total.functions++;
          }

          // Prefixed keywords
          if (!matches(property, /(^|-)(transition(-property)?|animation(-name)?)$/)) {
            for (let k of value.matchAll(/(?<![-a-z])-[a-z]+-[a-z-]+(?=$|\\s|,|\\/)/g)) {
              incrementByKey(ret.keywords, k);
              ret.total.keywords++;
            }
          }
        });
      }

      // Prefixed media features
      if (rule.media) {
        let features = rule.media
                  .replace(/\\s+/g, "")
                  .match(/\\(-[a-z]+-[\\w-]+(?=[:\\)])/g);

        if (features) {
          features = features.map(s => s.slice(1));

          for (let feature of features) {
            incrementByKey(ret.media, feature);
            ret.total.media++;
          }
        }
      }
    });

    ret.total.total = sumObject(ret.total);

    for (let type in ret) {
      ret[type] = sortObject(ret[type]);
    }

    return ret;
  }


  let ast = JSON.parse(css);
  let stats = compute(ast).total;
  return Object.entries(stats).flatMap(([prop, freq]) => {
    return Array(freq).fill(prop);
  });
}
catch (e) {
  return [];
}
'''
OPTIONS (library="gs://httparchive/lib/css-utils.js");

SELECT
  *
FROM (
  SELECT
    client,
    prop,
    COUNT(DISTINCT page) AS pages,
    COUNT(0) AS freq,
    SUM(COUNT(0)) OVER (PARTITION BY client) AS total,
    COUNT(0) / SUM(COUNT(0)) OVER (PARTITION BY client) AS pct
  FROM
    `httparchive.almanac.parsed_css`,
    UNNEST(getPrefixStats(css)) AS prop
  WHERE
    date = '2020-08-01'
  GROUP BY
    client,
    prop)
WHERE
  pages >= 1000
ORDER BY
  pct DESC
```

Reply:
Could you update the PR with the working query?